### PR TITLE
Feature/apl 1909 failed txs acceptance

### PR DESCRIPTION
**Pull request by AndrewBoyarsky**
- Add failed transactions support: transactions that are not passing state-dependent validation will be added as failed into a blockchain charging fee from the sender's account; specially configured transaction types may fail during execution and will be treated as failed. Failed transaction will be saved into db with a special `error_message` value, indicating the reason, why the transaction failed
- Add tests for __Blockchain Downloading__
- Add Failed Transaction p2p verification routine with 3 debug endpoints to verify/check failed transactions manually + tests
- Add failed transactions node/chain config parameters, updated blockchain config tests
- New exceptions hierarchy with a base exception: `AplCoreLogicException` to replace all checked exceptions inherited from AplException
- Introduce and document `AplUnacceptableTransactionValidationException` (state-independent check failed, transaction is not valid), `AplAcceptableTransactionValidationException` (state-dependent check failed, transaction may be accepted as failed), `AplTransactionExecutionException` for transaction, that may fail during execution with `TransactionType.canFailDuringExecution()=true`
- Refactor and remove boilerplate JDBC code by introducing `JdbcQueryExecutionHelper`

### CODE REVIEW

**Review by alzinchenko — APPROVED** [review]

**Review by yuriy-larin-firstbridge — APPROVED** [review]
Minor changes

**Review by andrii-firstbridge — APPROVED** [review]

File: apl-api/src/main/java/com/apollocurrency/aplwallet/api/p2p/request/GetNextBlocksRequest.java [hunk 1]
```diff
@@ -4,13 +4,15 @@
...
 
+import lombok.EqualsAndHashCode;
 import lombok.Getter;
 
 import java.util.List;
 import java.util.StringJoiner;
 import java.util.UUID;
 
 @Getter
+@EqualsAndHashCode
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-api/src/main/java/com/apollocurrency/aplwallet/api/p2p/request/GetNextBlocksRequest.java:15]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/service/fulltext/FullTextSearchServiceTest.java [hunk 2]
```diff
@@ -81,7 +80,7 @@
         TaggedDataServiceImpl.class, FullTextSearchUpdater.class,
         GlobalSyncImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/service/fulltext/FullTextSearchServiceTest.java:83]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/service/blockchain/BlockchainTest.java [hunk 3]
```diff
@@ -128,7 +126,7 @@
         BlockEntityRowMapper.class, BlockEntityToModelConverter.class, BlockModelToEntityConverter.class, JdbiHandleFactory.class, JdbiConfiguration.class,
         TransactionIndexDao.class, DaoConfig.class, JdbiTransactionalInterceptor.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/service/blockchain/BlockchainTest.java:129]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/service/state/DGSObserverTest.java [hunk 4]
```diff
@@ -99,7 +96,7 @@
         GlobalSyncImpl.class,
         FullTextConfigImpl.class,
         DGSPublicFeedbackTable.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/service/state/DGSObserverTest.java:99]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/http/post/CreateTransactionHandlerTest.java [hunk 5]
```diff
@@ -56,7 +55,7 @@
     public WeldInitiator weld = WeldInitiator.from(DbProperties.class, NtpTime.class,
         PropertiesHolder.class,
         TransactionDaoImpl.class, TransactionProcessor.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/http/post/CreateTransactionHandlerTest.java:58]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingVoteTableTest.java [hunk 6]
```diff
@@ -91,7 +90,7 @@
         BlockEntityRowMapper.class, BlockEntityToModelConverter.class, BlockModelToEntityConverter.class,
         TransactionDaoImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingVoteTableTest.java:93]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingResultTableTest.java [hunk 7]
```diff
@@ -93,7 +92,7 @@
         FullTextConfigImpl.class,
         DerivedDbTablesRegistryImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingResultTableTest.java:95]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingPollVoterTableTest.java [hunk 8]
```diff
@@ -88,7 +87,7 @@
         FullTextConfigImpl.class,
         DerivedDbTablesRegistryImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingPollVoterTableTest.java:90]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingPollTableTest.java [hunk 9]
```diff
@@ -111,7 +111,7 @@
         TransactionBuilderFactory.class,
         FullTextConfigImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingPollTableTest.java:114]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingPollLinkedTransactionTest.java [hunk 10]
```diff
@@ -82,7 +81,7 @@
         DerivedDbTablesRegistryImpl.class,
         TransactionBuilderFactory.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/phasing/PhasingPollLinkedTransactionTest.java:84]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSTagTableTest.java [hunk 11]
```diff
@@ -94,7 +93,7 @@
         DGSTagTable.class,
         DerivedDbTablesRegistryImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSTagTableTest.java:96]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSPurchaseTableTest.java [hunk 12]
```diff
@@ -89,7 +88,7 @@
         FullTextConfigImpl.class,
         DGSPurchaseTable.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSPurchaseTableTest.java:91]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSPublicFeedbackTableTest.java [hunk 13]
```diff
@@ -84,7 +83,7 @@
         BlockEntityRowMapper.class, BlockEntityToModelConverter.class, BlockModelToEntityConverter.class,
         TransactionDaoImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TransactionModelToEntityConverter.class, TransactionEntityToModelConverter.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TransactionModelToEntityConverter.class, TransactionEntityToModelConverter.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSPublicFeedbackTableTest.java:86]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSGoodsTableTest.java [hunk 14]
```diff
@@ -88,7 +87,7 @@
         GlobalSyncImpl.class,
         FullTextConfigImpl.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSGoodsTableTest.java:90]: done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSFeedbackTableTest.java [hunk 15]
```diff
@@ -76,7 +75,7 @@
         FullTextConfigImpl.class,
         DGSFeedbackTable.class,
         TransactionServiceImpl.class, ShardDbExplorerImpl.class,
-        TransactionRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
+        TransactionEntityRowMapper.class, TransactionEntityRowMapper.class, TxReceiptRowMapper.class, PrunableTxRowMapper.class,
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/dao/state/dgs/DGSFeedbackTableTest.java:78]: done

**Review by alzinchenko — APPROVED** [review]

**Review by alzinchenko — APPROVED** [review]

**Review by alzinchenko — APPROVED** [review]

**Pull request merged**